dpnp/linalg/dpnp_utils_linalg.py

```diff
@@ -907,6 +951,318 @@ def dpnp_inv(a):
+    )
+
+    q = q[..., :mc, :].swapaxes(-2, -1)
+    r = a_t[..., :mc].swapaxes(-2, -1)
+
+    r_triu = dpnp.triu(r).astype(res_type, copy=False)
+
+    q_reshape = q.reshape(batch_shape + q.shape[-2:])
+    r_triu_reshape = r_triu.reshape(batch_shape + r_triu.shape[-2:])
+
+    return (
+        q_reshape.astype(dtype=res_type, copy=False),
+        r_triu_reshape.astype(dtype=res_type, copy=False),
+    )
+
+
+def dpnp_qr(a, mode="reduced"):
+    """
+    dpnp_qr(a, mode="reduced")
+
+    Return the qr factorization of `a` matrix.
+
+    """
+
+    if a.ndim > 2:
+        return dpnp_qr_batch(a, mode=mode)
+
+    a_usm_arr = dpnp.get_usm_ndarray(a)
+    a_sycl_queue = a.sycl_queue
+    a_usm_type = a.usm_type
+
+    res_type = _common_type(a)
+
+    m, n = a.shape
+    k = min(m, n)
+    if k == 0:
+        if mode == "reduced":
+            return dpnp.empty(
+                (m, 0),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            ), dpnp.empty(
+                (0, n),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            )
+        elif mode == "complete":
+            return dpnp.identity(
+                m, dtype=res_type, sycl_queue=a_sycl_queue, usm_type=a_usm_type
+            ), dpnp.empty(
+                (m, n),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            )
+        elif mode == "r":
+            return dpnp.empty(
+                (0, n),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            )
+        else:  # mode == "raw"
+            return dpnp.empty(
+                (n, m),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            ), dpnp.empty(
+                (0,),
+                dtype=res_type,
+                sycl_queue=a_sycl_queue,
+                usm_type=a_usm_type,
+            )
+
+    a = a.T
+    a_usm_arr = dpnp.get_usm_ndarray(a)
+    a_t = dpnp.empty_like(a, order="C", dtype=res_type)
+
+    # use DPCTL tensor function to fill the matrix array
+    # with content from the input array `a`
+    a_ht_copy_ev, a_copy_ev = ti._copy_usm_ndarray_into_usm_ndarray(
+        src=a_usm_arr, dst=a_t.get_array(), sycl_queue=a_sycl_queue
+    )
+
+    tau_h = dpnp.empty(
+        k, dtype=res_type, sycl_queue=a_sycl_queue, usm_type=a_usm_type
+    )
+
+    # Call the LAPACK extension function _geqrf to compute the QR factorization
+    # of a general m x n matrix.
+    ht_geqrf_ev, geqrf_ev = li._geqrf(
+        a_sycl_queue, a_t.get_array(), tau_h.get_array(), [a_copy_ev]
+    )
+
+    ht_geqrf_ev.wait()
+    a_ht_copy_ev.wait()
+
+    if mode == "r":
+        r = a_t[:, :k].transpose()
+        return dpnp.triu(r).astype(res_type, copy=False)
```

Comment:
will it make sense to rewrite like below? or there is no performance gain of implementing that?
```python
    r = a_t[:, :k].transpose()
    host_tasks_list = [ht_geqrf_ev, a_ht_copy_ev]
    r = _triu_inplace(r, host_tasks_list, [geqrf_ev ])
    dpctl.SyclEvent.wait_for(host_tasks_list)
    return r

...

def _triu_inplace(a, host_tasks, depends=[]):
    # TODO: implement a dedicated kernel for in-place triu instead of extra memory allocation for result
    out = dpnp.empty_like(a, order="C")
    ht_triu_ev, _ = ti._triu(src=a.get_array(), dst=out.get_array(), k=0, sycl_queue=a.sycl_queue, depends=depends)
    host_tasks.append(ht_triu_ev)
    return out
```

Reply:
Done
Thanks